File: rust/src/connection.rs
// Copyright (c) 2025 ADBC Drivers Contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Connection implementation for the Databricks ADBC driver.

use crate::client::DatabricksClient;
use crate::error::DatabricksErrorHelper;
use crate::metadata::builder::{
    build_get_objects_all, build_get_objects_catalogs, build_get_objects_schemas,
    build_get_objects_tables, collect_reader, filter_by_pattern, group_schemas_by_catalog,
    group_tables_and_columns, group_tables_by_catalog_schema,
};
use crate::metadata::parse::{
    parse_catalogs, parse_columns, parse_columns_as_fields, parse_schemas, parse_tables,
};
use crate::statement::Statement;
use adbc_core::error::Result;
use adbc_core::options::{InfoCode, ObjectDepth, OptionConnection, OptionValue};
use adbc_core::schemas::GET_TABLE_TYPES_SCHEMA;
use adbc_core::Optionable;
use arrow_array::{RecordBatch, RecordBatchIterator, RecordBatchReader, StringArray};
use arrow_schema::{ArrowError, Schema};
use driverbase::error::ErrorHelper;
use std::collections::{HashMap, HashSet};
use std::sync::Arc;
use tracing::span::EnteredSpan;
use tracing::{debug, info_span, Span};

/// Configuration passed from Database to Connection.
pub struct ConnectionConfig {
    pub host: String,
    pub warehouse_id: String,
    pub catalog: Option<String>,
    pub schema: Option<String>,
    pub client: Arc<dyn DatabricksClient>,
}

/// Represents an active connection to a Databricks SQL endpoint.
///
/// A Connection is created from a Database and is used to create Statements
/// for executing SQL queries. It maintains a session with the Databricks
/// server and manages shared resources like the HTTP client.
#[derive(Debug)]
pub struct Connection {
    // Configuration
    host: String,
    warehouse_id: String,

    // Databricks client (trait object for backend flexibility)
    client: Arc<dyn DatabricksClient>,

    // Session ID (created on connection initialization)
    session_id: String,

    // Tokio runtime for async operations
    runtime: tokio::runtime::Runtime,

    // Tracing span that attaches session_id to all log lines within this connection
    _log_span: EnteredSpan,
}

/// Type alias for our empty reader used in stub implementations.
type EmptyReader =
    RecordBatchIterator<std::vec::IntoIter<std::result::Result<RecordBatch, ArrowError>>>;

impl Connection {
    /// Called by Database::new_connection().
    ///
    /// Connection receives the DatabricksClient and runtime from Database.
    /// The runtime is created by Database so it can share the handle with
    /// SeaClient and ResultReaderFactory before Connection is created.
    pub(crate) fn new_with_runtime(
        config: ConnectionConfig,
        runtime: tokio::runtime::Runtime,
    ) -> crate::error::Result<Self> {
        // Enter the ADBC span before any work so all log lines are tagged.
        // session_id is recorded once the session is created.
        let span = info_span!("ADBC", session_id = tracing::field::Empty);
        let entered = span.entered();

        debug!(
            "Creating connection to {} with warehouse {}",
            config.host, config.warehouse_id
        );

        // Create session using the client provided by Database
        let session_info = runtime.block_on(config.client.create_session(
            config.catalog.as_deref(),
            config.schema.as_deref(),
            HashMap::new(),
        ))?;

        Span::current().record("session_id", session_info.session_id.as_str());

        Ok(Self {
            host: config.host,
            warehouse_id: config.warehouse_id,
            client: config.client,
            session_id: session_info.session_id,
            runtime,
            _log_span: entered,
        })
    }

    /// Returns the Databricks host URL.
    pub fn host(&self) -> &str {
        &self.host
    }

    /// Returns the warehouse ID.
    pub fn warehouse_id(&self) -> &str {
        &self.warehouse_id
    }

    /// Returns the session ID.
    pub fn session_id(&self) -> &str {
        &self.session_id
    }
}

impl Optionable for Connection {
    type Option = OptionConnection;

    fn set_option(&mut self, key: Self::Option, _value: OptionValue) -> Result<()> {
        match key {
            OptionConnection::AutoCommit => {
                // Databricks SQL doesn't support transactions in the traditional sense
                // Just accept and ignore this option
                Ok(())
            }
            _ => Err(DatabricksErrorHelper::set_unknown_option(&key).to_adbc()),
        }
    }

    fn get_option_string(&self, key: Self::Option) -> Result<String> {
        Err(DatabricksErrorHelper::get_unknown_option(&key).to_adbc())
    }

    fn get_option_bytes(&self, key: Self::Option) -> Result<Vec<u8>> {
        Err(DatabricksErrorHelper::get_unknown_option(&key).to_adbc())
    }

    fn get_option_int(&self, key: Self::Option) -> Result<i64> {
        Err(DatabricksErrorHelper::get_unknown_option(&key).to_adbc())
    }

    fn get_option_double(&self, key: Self::Option) -> Result<f64> {
        Err(DatabricksErrorHelper::get_unknown_option(&key).to_adbc())
    }
}

impl adbc_core::Connection for Connection {
    type StatementType = Statement;

    fn new_statement(&mut self) -> Result<Self::StatementType> {
        Ok(Statement::new(
            self.client.clone(),
            self.session_id.clone(),
            self.runtime.handle().clone(),
        ))
    }

    fn cancel(&mut self) -> Result<()> {
        // TODO: Implement connection-level cancellation
        Ok(())
    }

    fn get_info(&self, codes: Option<HashSet<InfoCode>>) -> Result<impl RecordBatchReader + Send> {
        use driverbase::InfoBuilder;

        let mut builder = InfoBuilder::new();

        // Filter by requested codes or return all if none specified
        let return_all = codes.is_none();
        let codes = codes.unwrap_or_default();

        if return_all || codes.contains(&InfoCode::DriverName) {
            builder.add_string(u32::from(&InfoCode::DriverName), "Databricks ADBC Driver");
        }
        if return_all || codes.contains(&InfoCode::DriverVersion) {
            builder.add_string(
                u32::from(&InfoCode::DriverVersion),
                env!("CARGO_PKG_VERSION"),
            );
        }
        if return_all || codes.contains(&InfoCode::VendorName) {
            builder.add_string(u32::from(&InfoCode::VendorName), "Databricks");
        }

        Ok(builder.build())
    }

    fn get_objects(
        &self,
        depth: ObjectDepth,
        catalog: Option<&str>,
        db_schema: Option<&str>,
        table_name: Option<&str>,
        table_type: Option<Vec<&str>>,
        column_name: Option<&str>,
    ) -> Result<impl RecordBatchReader + Send> {
        debug!(
            "get_objects: depth={:?}, catalog={:?}, db_schema={:?}, table_name={:?}, table_type={:?}, column_name={:?}",
            depth, catalog, db_schema, table_name, table_type, column_name
        );
        match depth {
            ObjectDepth::Catalogs => {
                let result = self
                    .runtime
                    .block_on(self.client.list_catalogs(&self.session_id))
                    .map_err(|e| e.to_adbc())?;
                let catalogs = parse_catalogs(result).map_err(|e| e.to_adbc())?;

                // Client-side catalog pattern filtering (SHOW CATALOGS has no LIKE clause)
                let catalogs = filter_by_pattern(catalogs, catalog, |c| &c.catalog_name);
                debug!("get_objects(Catalogs): found {} catalogs", catalogs.len());

                let reader = build_get_objects_catalogs(catalogs).map_err(|e| e.to_adbc())?;
                collect_reader(reader).map_err(|e| e.to_adbc())
            }

            ObjectDepth::Schemas => {
                let result = self
                    .runtime
                    .block_on(
                        self.client
                            .list_schemas(&self.session_id, catalog, db_schema),
                    )
                    .map_err(|e| e.to_adbc())?;
                let schemas = parse_schemas(result, catalog).map_err(|e| e.to_adbc())?;
                debug!("get_objects(Schemas): found {} schemas", schemas.len());

                let grouped = group_schemas_by_catalog(schemas);
                let reader = build_get_objects_schemas(grouped).map_err(|e| e.to_adbc())?;
                collect_reader(reader).map_err(|e| e.to_adbc())
            }

            ObjectDepth::Tables => {
                let result = self
                    .runtime
                    .block_on(self.client.list_tables(
                        &self.session_id,
                        catalog,
                        db_schema,
                        table_name,
                        table_type.as_deref(),
                    ))
                    .map_err(|e| e.to_adbc())?;
                let mut tables = parse_tables(result).map_err(|e| e.to_adbc())?;

                // Client-side table type filtering
                if let Some(ref types) = table_type {
                    tables.retain(|t| types.iter().any(|tt| t.table_type.eq_ignore_ascii_case(tt)));
                }
                debug!(
                    "get_objects(Tables): found {} tables after filtering",
                    tables.len()
                );

                let grouped = group_tables_by_catalog_schema(tables);
                let reader = build_get_objects_tables(grouped).map_err(|e| e.to_adbc())?;
                collect_reader(reader).map_err(|e| e.to_adbc())
            }

            ObjectDepth::Columns | ObjectDepth::All => {
                // Step 1: Get tables (for table_type info)
                let tables_result = self
                    .runtime
                    .block_on(self.client.list_tables(
                        &self.session_id,
                        catalog,
                        db_schema,
                        table_name,
                        table_type.as_deref(),
                    ))
                    .map_err(|e| e.to_adbc())?;
                let mut tables = parse_tables(tables_result).map_err(|e| e.to_adbc())?;

                // Client-side table type filtering
                if let Some(ref types) = table_type {
                    tables.retain(|t| types.iter().any(|tt| t.table_type.eq_ignore_ascii_case(tt)));
                }
                debug!(
                    "get_objects(All): found {} tables after filtering",
                    tables.len()
                );

                // Step 2: Get distinct catalogs from tables result
                let distinct_catalogs: Vec<String> = {
                    let mut cats: Vec<String> = tables
                        .iter()
                        .map(|t| t.catalog_name.clone())
                        .collect::<HashSet<_>>()
                        .into_iter()
                        .collect();
                    cats.sort();
                    cats
                };

                // Step 3: Fetch columns per catalog (parallel)
                debug!(
                    "get_objects(All): fetching columns for {} distinct catalogs: {:?}",
                    distinct_catalogs.len(),
                    distinct_catalogs
                );
                let all_columns = self
                    .runtime
                    .block_on(async {
                        let mut handles = Vec::new();
                        for cat in &distinct_catalogs {
                            let client = self.client.clone();
                            let session_id = self.session_id.clone();
                            let cat = cat.clone();
                            let schema_pattern = db_schema.map(|s| s.to_string());
                            let table_pattern = table_name.map(|s| s.to_string());
                            let col_pattern = column_name.map(|s| s.to_string());

                            handles.push(tokio::spawn(async move {
                                client
                                    .list_columns(
                                        &session_id,
                                        &cat,
                                        schema_pattern.as_deref(),
                                        table_pattern.as_deref(),
                                        col_pattern.as_deref(),
                                    )
                                    .await
                            }));
                        }

                        let mut all_columns = Vec::new();
                        for handle in handles {
                            let result = handle.await.map_err(|e| {
                                crate::error::DatabricksErrorHelper::io()
                                    .message(format!("Column fetch task failed: {}", e))
                            })?;
                            let cols = parse_columns(result?)?;
                            all_columns.extend(cols);
                        }
                        Ok::<_, crate::error::Error>(all_columns)
                    })
                    .map_err(|e| e.to_adbc())?;

                // Step 4: Group and build
                debug!(
                    "get_objects(All): fetched {} total columns",
                    all_columns.len()
                );
                let grouped = group_tables_and_columns(tables, all_columns);
                let reader = build_get_objects_all(grouped).map_err(|e| e.to_adbc())?;
                collect_reader(reader).map_err(|e| e.to_adbc())
            }
        }
    }

    fn get_table_schema(
        &self,
        catalog: Option<&str>,
        db_schema: Option<&str>,
        table_name: &str,
    ) -> Result<Schema> {
        debug!(
            "get_table_schema: catalog={:?}, db_schema={:?}, table_name={:?}",
            catalog, db_schema, table_name
        );
        // SHOW COLUMNS IN CATALOG `{cat}` requires a catalog.
        // If catalog is not provided, discover it via list_tables.
        let resolved_catalog = match catalog {
            Some(c) if !c.is_empty() => c.to_string(),
            _ => {
                let result = self
                    .runtime
                    .block_on(self.client.list_tables(
                        &self.session_id,
                        None,
                        db_schema,
                        Some(table_name),
                        None,
                    ))
                    .map_err(|e| e.to_adbc())?;
                let tables = parse_tables(result).map_err(|e| e.to_adbc())?;
                tables
                    .first()
                    .map(|t| t.catalog_name.clone())
                    .ok_or_else(|| {
                        DatabricksErrorHelper::not_found()
                            .message(format!("Table not found: {}", table_name))
                            .to_adbc()
                    })?
            }
        };

        debug!("get_table_schema: resolved catalog={:?}", resolved_catalog);

        let result = self
            .runtime
            .block_on(self.client.list_columns(
                &self.session_id,
                &resolved_catalog,
                db_schema,
                Some(table_name),
                None, // all columns
            ))
            .map_err(|e| e.to_adbc())?;
        let fields = parse_columns_as_fields(result).map_err(|e| e.to_adbc())?;

        if fields.is_empty() {
            return Err(DatabricksErrorHelper::not_found()
                .message(format!("Table not found: {}", table_name))
                .to_adbc());
        }

        debug!(
            "get_table_schema: found {} fields for {}.{}",
            fields.len(),
            resolved_catalog,
            table_name
        );
        Ok(Schema::new(fields))
    }

    fn get_table_types(&self) -> Result<impl RecordBatchReader + Send> {
        let table_types = self.client.list_table_types();

        let array = StringArray::from(table_types);
        let batch = RecordBatch::try_new(GET_TABLE_TYPES_SCHEMA.clone(), vec![Arc::new(array)])
            .map_err(|e| {
                DatabricksErrorHelper::io()
                    .message(format!("Failed to build table types result: {}", e))
                    .to_adbc()
            })?;

        Ok(RecordBatchIterator::new(
            vec![Ok(batch)].into_iter(),
            GET_TABLE_TYPES_SCHEMA.clone(),
        ))
    }

    fn read_partition(
        &self,
        _partition: impl AsRef<[u8]>,
    ) -> Result<impl RecordBatchReader + Send> {
        Err::<EmptyReader, _>(
            DatabricksErrorHelper::not_implemented()
                .message("read_partition")
                .to_adbc(),
        )
    }

    fn commit(&mut self) -> Result<()> {
        // Databricks SQL is auto-commit only
        Ok(())
    }

    fn rollback(&mut self) -> Result<()> {
        // Databricks SQL doesn't support rollback
        Err(DatabricksErrorHelper::not_implemented()
            .message("rollback - Databricks SQL is auto-commit only")
            .to_adbc())
    }

    fn get_statistic_names(&self) -> Result<impl RecordBatchReader + Send> {
        Err::<EmptyReader, _>(
            DatabricksErrorHelper::not_implemented()
                .message("get_statistic_names")
                .to_adbc(),
        )
    }

    fn get_statistics(
        &self,
        _catalog: Option<&str>,
        _db_schema: Option<&str>,
        _table_name: Option<&str>,
        _approximate: bool,
    ) -> Result<impl RecordBatchReader + Send> {
        Err::<EmptyReader, _>(
            DatabricksErrorHelper::not_implemented()
                .message("get_statistics")
                .to_adbc(),
        )
    }
}

impl Drop for Connection {
    fn drop(&mut self) {
        // Clean up session on connection close
        debug!("Closing session: {}", self.session_id);
        let _ = self
            .runtime
            .block_on(self.client.delete_session(&self.session_id));
    }
}

#[cfg(test)]
mod tests {
    // Note: Full connection tests require mock DatabricksClient
    // Integration tests should be added in a separate test module
}

File: go/connection.go
// Copyright (c) 2025 ADBC Drivers Contributors
//
// This file has been modified from its original version, which is
// under the Apache License:
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package databricks

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"strings"

	"github.com/adbc-drivers/driverbase-go/driverbase"
	"github.com/apache/arrow-adbc/go/adbc"
	_ "github.com/databricks/databricks-sql-go"
	dbsqlerr "github.com/databricks/databricks-sql-go/errors"
)

type connectionImpl struct {
	driverbase.ConnectionImplBase

	// Connection settings
	catalog  string
	dbSchema string

	// Database connection
	conn *sql.Conn

	// Arrow serialization options
	useArrowNativeGeospatial bool
}

func (c *connectionImpl) Close() error {
	if c.conn == nil {
		return adbc.Error{Code: adbc.StatusInvalidState}
	}
	defer func() {
		c.conn = nil
	}()
	return c.conn.Close()
}

func (c *connectionImpl) NewStatement() (adbc.Statement, error) {
	return &statementImpl{
		StatementImplBase: driverbase.NewStatementImplBase(&c.ConnectionImplBase, c.ErrorHelper),
		conn:              c,
		bulkIngestOptions: driverbase.NewBulkIngestOptions(),
	}, nil
}

func (c *connectionImpl) SetAutocommit(autocommit bool) error {
	// Databricks SQL doesn't support explicit transaction control in the same way
	// as traditional databases. Most operations are implicitly committed.
	if !autocommit {
		return adbc.Error{
			Code: adbc.StatusNotImplemented,
			Msg:  "disabling autocommit is not supported",
		}
	}
	return nil
}

// CurrentNamespacer interface implementation
func (c *connectionImpl) GetCurrentCatalog() (string, error) {
	if c.catalog != "" {
		return c.catalog, nil
	}

	if c.conn == nil {
		return "", adbc.Error{
			Code: adbc.StatusInvalidState,
			Msg:  "connection is nil",
		}
	}

	var catalog string
	err := c.conn.QueryRowContext(context.Background(), "SELECT current_catalog()").Scan(&catalog)
	if err != nil {
		return "", adbc.Error{
			Code: adbc.StatusInternal,
			Msg:  fmt.Sprintf("failed to get current catalog: %v", err),
		}
	}

	return catalog, nil
}

func (c *connectionImpl) GetCurrentDbSchema() (string, error) {
	if c.dbSchema != "" {
		return c.dbSchema, nil
	}

	if c.conn == nil {
		return "", adbc.Error{
			Code: adbc.StatusInvalidState,
			Msg:  "connection is nil",
		}
	}

	var schema string
	err := c.conn.QueryRowContext(context.Background(), "SELECT current_schema()").Scan(&schema)
	if err != nil {
		return "", adbc.Error{
			Code: adbc.StatusInternal,
			Msg:  fmt.Sprintf("failed to get current schema: %v", err),
		}
	}

	return schema, nil
}

func (c *connectionImpl) SetCurrentCatalog(catalog string) error {
	if catalog == "" {
		return adbc.Error{
			Code: adbc.StatusInvalidArgument,
			Msg:  "catalog cannot be empty",
		}
	}
	if c.conn == nil {
		return adbc.Error{
			Code: adbc.StatusInvalidState,
			Msg:  "failed to set catalog: connection is nil",
		}
	}
	escapedCatalog := strings.ReplaceAll(catalog, "`", "``")
	_, err := c.conn.ExecContext(context.Background(), fmt.Sprintf("USE CATALOG `%s`", escapedCatalog))
	if err != nil {
		return adbc.Error{
			Code: adbc.StatusInternal,
			Msg:  fmt.Sprintf("failed to set catalog: %v", err),
		}
	}
	c.catalog = catalog
	return nil
}

func (c *connectionImpl) SetCurrentDbSchema(schema string) error {
	if schema == "" {
		return adbc.Error{
			Code: adbc.StatusInvalidArgument,
			Msg:  "schema cannot be empty",
		}
	}
	if c.conn == nil {
		return adbc.Error{
			Code: adbc.StatusInvalidState,
			Msg:  "failed to set db schema: connection is nil",
		}
	}
	escapedSchema := strings.ReplaceAll(schema, "`", "``")
	_, err := c.conn.ExecContext(context.Background(), fmt.Sprintf("USE SCHEMA `%s`", escapedSchema))
	if err != nil {
		return adbc.Error{
			Code: adbc.StatusInternal,
			Msg:  fmt.Sprintf("failed to set schema: %v", err),
		}
	}
	c.dbSchema = schema
	return nil
}

// TableTypeLister interface implementation
func (c *connectionImpl) ListTableTypes(ctx context.Context) ([]string, error) {
	// Databricks supports these table types
	return []string{"TABLE", "VIEW", "EXTERNAL_TABLE", "MANAGED_TABLE", "STREAMING_TABLE", "MATERIALIZED_VIEW"}, nil
}

// Transaction methods (Databricks has limited transaction support)
func (c *connectionImpl) Commit(ctx context.Context) error {
	// Most operations are auto-committed.
	return adbc.Error{
		Code: adbc.StatusNotImplemented,
		Msg:  "Commit is not supported",
	}
}

func (c *connectionImpl) Rollback(ctx context.Context) error {
	// Databricks SQL doesn't support explicit transactions in the traditional sense.
	// Most operations are auto-committed.
	return adbc.Error{
		Code: adbc.StatusNotImplemented,
		Msg:  "rollback is not supported",
	}
}

// DbObjectsEnumerator interface implementation
func (c *connectionImpl) GetCatalogs(ctx context.Context, catalogFilter *string) (catalogs []string, err error) {
	catalogs = []string{}
	query := "SHOW CATALOGS"
	if catalogFilter != nil {
		escapedFilter := strings.ReplaceAll(*catalogFilter, "'", "''")
		query += fmt.Sprintf(" LIKE '%s'", escapedFilter)
	}
	var rows *sql.Rows
	rows, err = c.conn.QueryContext(ctx, query)
	if err != nil {
		return nil, adbc.Error{
			Code: adbc.StatusInternal,
			Msg:  fmt.Sprintf("failed to query catalogs: %v", err),
		}
	}
	defer func() {
		err = errors.Join(err, rows.Close())
	}()

	for rows.Next() {
		var catalog string
		if err := rows.Scan(&catalog); err != nil {
			return nil, adbc.Error{
				Code: adbc.StatusInternal,
				Msg:  fmt.Sprintf("failed to scan catalog: %v", err),
			}
		}
		catalogs = append(catalogs, catalog)
	}

	return catalogs, errors.Join(err, rows.Err())
}

func (c *connectionImpl) GetDBSchemasForCatalog(ctx context.Context, catalog string, schemaFilter *string) (schemas []string, err error) {
	schemas = []string{}
	escapedCatalog := strings.ReplaceAll(catalog, "`", "``")
	query := fmt.Sprintf("SHOW SCHEMAS IN `%s`", escapedCatalog)
	if schemaFilter != nil {
		escapedFilter := strings.ReplaceAll(*schemaFilter, "'", "''")
		query += fmt.Sprintf(" LIKE '%s'", escapedFilter)
	}

	var rows *sql.Rows
	rows, err = c.conn.QueryContext(ctx, query)
	if err != nil {
		return nil, adbc.Error{
			Code: adbc.StatusInternal,
			Msg:  fmt.Sprintf("failed to query schemas: %v", err),
		}
	}
	defer func() {
		err = errors.Join(err, rows.Close())
	}()
	for rows.Next() {
		var schema string
		if err := rows.Scan(&schema); err != nil {
			return nil, adbc.Error{
				Code: adbc.StatusInternal,
				Msg:  fmt.Sprintf("failed to scan schema: %v", err),
			}
		}
		schemas = append(schemas, schema)
	}

	err = errors.Join(err, rows.Err())
	return schemas, err
}

func (c *connectionImpl) GetTablesForDBSchema(ctx context.Context, catalog string, schema string, tableFilter *string, columnFilter *string, includeColumns bool) (tables []driverbase.TableInfo, err error) {
	if includeColumns {
		return c.getTablesWithColumns(ctx, catalog, schema, tableFilter, columnFilter)
	}

	tables = []driverbase.TableInfo{}
	escapedCatalog := strings.ReplaceAll(catalog, "`", "``")
	escapedSchema := strings.ReplaceAll(schema, "`", "``")
	query := fmt.Sprintf("SHOW TABLES IN `%s`.`%s`", escapedCatalog, escapedSchema)
	if tableFilter != nil {
		escapedFilter := strings.ReplaceAll(*tableFilter, "'", "''")
		query += fmt.Sprintf(" LIKE '%s'", escapedFilter)
	}

	var rows *sql.Rows
	rows, err = c.conn.QueryContext(ctx, query)
	if err != nil {
		return nil, adbc.Error{
			Code: adbc.StatusInternal,
			Msg:  fmt.Sprintf("failed to query tables: %v", err),
		}
	}
	defer func() {
		err = errors.Join(err, rows.Close())
	}()
	for rows.Next() {
		var database, tableName, isTemporary string
		if err := rows.Scan(&database, &tableName, &isTemporary); err != nil {
			return nil, adbc.Error{
				Code: adbc.StatusInternal,
				Msg:  fmt.Sprintf("failed to scan table: %v", err),
			}
		}

		tableInfo := driverbase.TableInfo{
			TableName:        tableName,
			TableType:        "TABLE", // Default to TABLE, could be improved with more detailed queries
			TableColumns:     []driverbase.ColumnInfo{},
			TableConstraints: []driverbase.ConstraintInfo{},
		}

		tables = append(tables, tableInfo)
	}

	return tables, errors.Join(err, rows.Err())
}

// getTablesWithColumns retrieves complete table and column information using INFORMATION_SCHEMA
func (c *connectionImpl) getTablesWithColumns(ctx context.Context, catalog string, schema string, tableFilter *string, columnFilter *string) (tables []driverbase.TableInfo, err error) {
	tables = []driverbase.TableInfo{}

	lowerCatalog := strings.ToLower(catalog)

	// Skip internal catalogs that do not support metadata queries
	if lowerCatalog == "__databricks_internal" {
		return tables, nil
	}

	var queryBuilder strings.Builder
	queryBuilder.WriteString("SELECT DISTINCT c.TABLE_NAME, c.ordinal_position, c.COLUMN_NAME, c.DATA_TYPE, c.IS_NULLABLE FROM ")
	if lowerCatalog == "hive_metastore" || lowerCatalog == "system" {
		// Hive Metastore and system catalog metadata are only available via the system-level information_schema
		queryBuilder.WriteString("system.information_schema.COLUMNS c ")
		queryBuilder.WriteString("WHERE c.table_catalog = ")
		queryBuilder.WriteString(quoteString(catalog))
		queryBuilder.WriteString(" AND c.TABLE_SCHEMA = ")
		queryBuilder.WriteString(quoteString(schema))
	} else {
		// Unity Catalog catalogs have their own information_schema
		queryBuilder.WriteString(quoteIdentifier(catalog))
		queryBuilder.WriteString(".information_schema.COLUMNS c WHERE c.TABLE_SCHEMA = ")
		queryBuilder.WriteString(quoteString(schema))
	}

	if tableFilter != nil {
		queryBuilder.WriteString(" AND c.TABLE_NAME LIKE ")
		queryBuilder.WriteString(quoteString(*tableFilter))
	}
	if columnFilter != nil {
		queryBuilder.WriteString(" AND c.COLUMN_NAME LIKE ")
		queryBuilder.WriteString(quoteString(*columnFilter))
	}

	queryBuilder.WriteString(" ORDER BY c.TABLE_NAME, c.ordinal_position")

	rows, err := c.conn.QueryContext(ctx, queryBuilder.String())
	if err != nil {
		// If we don't have permissions on the catalog, this will
		// error. Catch that and simply return no tables instead of
		// blowing up.
		var dbExecutionErr dbsqlerr.DBExecutionError
		if errors.As(err, &dbExecutionErr) && dbExecutionErr.SqlState() == "42501" {
			return tables, nil
		}
		return nil, adbc.Error{
			Code: adbc.StatusInternal,
			Msg:  fmt.Sprintf("failed to query tables with columns: %v", err),
		}
	}
	defer func() {
		err = errors.Join(err, rows.Close())
	}()

	var currentTable *driverbase.TableInfo

	for rows.Next() {
		var tableName, columnName, dataType, isNullable string
		var ordinalPosition sql.NullInt32

		if err := rows.Scan(
			&tableName,
			&ordinalPosition, &columnName,
			&dataType, &isNullable,
		); err != nil {
			return nil, adbc.Error{
				Code: adbc.StatusInternal,
				Msg:  fmt.Sprintf("failed to scan table with columns: %v", err),
			}
		}

		// Check if we need to create a new table entry
		if currentTable == nil || currentTable.TableName != tableName {
			tables = append(tables, driverbase.TableInfo{
				TableName:        tableName,
				TableType:        "TABLE",
				TableColumns:     []driverbase.ColumnInfo{},
				TableConstraints: []driverbase.ConstraintInfo{},
			})
			currentTable = &tables[len(tables)-1]
		}

		var nullable *int16
		var isNullablePtr *string
		switch isNullable {
		case "YES":
			n := int16(driverbase.XdbcColumnNullable)
			nullable = &n
			isNullablePtr = &isNullable
		case "NO":
			n := int16(driverbase.XdbcColumnNoNulls)
			nullable = &n
			isNullablePtr = &isNullable
		}

		columnInfo := driverbase.ColumnInfo{
			ColumnName:     columnName,
			XdbcTypeName:   &dataType,
			XdbcNullable:   nullable,
			XdbcIsNullable: isNullablePtr,
		}

		if ordinalPosition.Valid {
			// Databricks uses 0-based indexing
			pos := ordinalPosition.Int32 + 1
			columnInfo.OrdinalPosition = &pos
		}

		currentTable.TableColumns = append(currentTable.TableColumns, columnInfo)
	}

	return tables, errors.Join(err, rows.Err())
}

// PrepareDriverInfo implements driverbase.DriverInfoPreparer.
func (c *connectionImpl) PrepareDriverInfo(ctx context.Context, infoCodes []adbc.InfoCode) error {
	var versionJSON string
	err := c.conn.QueryRowContext(ctx, "SELECT current_version()").Scan(&versionJSON)
	if err != nil {
		return adbc.Error{
			Code: adbc.StatusInternal,
			Msg:  fmt.Sprintf("failed to get vendor version: %v", err),
		}
	}

	var versionData map[string]any
	if err := json.Unmarshal([]byte(versionJSON), &versionData); err != nil {
		return c.DriverInfo.RegisterInfoCode(adbc.InfoVendorVersion, "unknown")
	}

	version := "unknown"
	if dbsqlVersion, ok := versionData["dbsql_version"].(string); ok && dbsqlVersion != "" {
		version = dbsqlVersion
	} else if dbrVersion, ok := versionData["dbr_version"].(string); ok {
		version = dbrVersion
	}

	return c.DriverInfo.RegisterInfoCode(adbc.InfoVendorVersion, version)
}

// quoteString escapes string literals using single quotes
func quoteString(value string) string {
	return fmt.Sprintf("'%s'", strings.ReplaceAll(value, "'", "''"))
}
